src/general/MainMenu.cs

```diff
@@ -100,6 +100,17 @@ public void SetCurrentMenu(uint index, bool slide = true)
         }
     }
 
+    [RunOnKeyDown("ui_cancel")]
+    public void ReturnPressed()
+    {
+        if (options.Visible)
```

Comment:
I think these should directly call `OnReturnFromOptions` etc. to not rely on the concrete options etc. classes to reduce coupling.

Reply:
No when returning from Options it checks if the settings need saving. If directly call OnReturnFromOptions then the settings won't be saved and you'll not get informed.

![image](https://user-images.githubusercontent.com/60734649/123421852-fd63ac80-d5ef-11eb-8ffc-27d56d061c09.png)

Reply:
Okay that's fine, but the other menus don't need anything special to happen when closing them, right?

Reply:
Still think we should use its own logic first to avoid future problems. (What if one day there's something to check?

Reply:
At that point, I think the ESC key handlers should be put in the separate menu classes, and the main menu would only handle like the tools menu.

Reply:
Some time ago you said that main menu should handle these ESC keystrikes because ESC also cause pause menu in the game exit.

Reply:
And it does.

Reply:
It's hard to balance the design:
- Optimally objects shouldn't care about other's internal state and mess with them
- The pause menu already exists with ESC so special handling is needed in the options menu if it were to handle ESC press (I think using a lower priority input handler would work)
- Some of the main menu sub menus don't have their own code so the main menu class needs to handle those